awsx/ec2/subnetDistributor.test.ts

```diff
@@ -12,8 +12,168 @@
+          azs: fc.array(fc.string({ size: "xsmall" }), { minLength: 2, maxLength: 4 }),
+          subnetSpecs: fc.array(subnetSpecNoMask(), { minLength: 1, maxLength: 5 }),
+        }),
+        ({ vpcCidrMask, azs, subnetSpecs }) => {
+          const vpcCidr = `10.0.0.0/${vpcCidrMask}`;
+
+          const result = getSubnetSpecs("vpcName", vpcCidr, azs, subnetSpecs);
+
+          for (const subnet of result) {
+            const subnetMask = getCidrMask(subnet.cidrBlock);
+            // Larger mask means smaller subnet
+            expect(subnetMask).toBeGreaterThan(vpcCidrMask);
+          }
+        },
+      ),
+    );
+  });
+
+  it("should use whole space if there's only one subnet and VPC is small", () => {
+    fc.assert(
+      fc.property(
+        fc.record({
+          vpcCidrMask: cidrMask({ min: 24 }),
+          subnetSpec: subnetSpecNoMask(),
+        }),
+        ({ vpcCidrMask, subnetSpec }) => {
+          const vpcCidr = `10.0.0.0/${vpcCidrMask}`;
+
+          const result = getSubnetSpecs("vpcName", vpcCidr, ["us-east-1a"], [subnetSpec]);
+
+          expect(result.length).toBe(1);
+          expect(result[0].cidrBlock).toBe(vpcCidr);
+        },
+      ),
+    );
+  });
+
+  it("should use default values if VPC is large", () => {
+    fc.assert(
+      fc.property(
+        fc.record({
+          vpcCidrMask: cidrMask({ max: 17 }),
+          subnetSpec: subnetSpecNoMask(),
+        }),
+        ({ vpcCidrMask }) => {
+          const vpcCidr = `10.0.0.0/${vpcCidrMask}`;
+
+          const result = getSubnetSpecs(
+            "vpcName",
+            vpcCidr,
+            ["us-east-1a"],
+            [{ type: "Public" }, { type: "Private" }, { type: "Isolated" }],
+          );
+
+          expect(result).toMatchObject([
+            {
+              cidrBlock: "10.0.0.0/19",
+              subnetName: "vpcName-private-1",
+              type: "Private",
+            },
+            {
+              cidrBlock: "10.0.32.0/20",
+              subnetName: "vpcName-public-1",
+              type: "Public",
+            },
+            {
+              cidrBlock: "10.0.48.0/24",
+              subnetName: "vpcName-isolated-1",
+              type: "Isolated",
+            },
+          ]);
+        },
+      ),
+    );
+  });
+
+  it("should not have overlapping ranges", () => {
+    fc.assert(
+      fc.property(
+        fc
+          .record({
+            vpcCidrMask: cidrMask(),
+            // We're just focusing on the logic of choosing the next range, so we don't need large numbers of specs.
+            // Similarly, we only need a single AZ because AZs are always split evenly.
+            subnetSpecs: fc.array(subnetSpec(), { minLength: 1, maxLength: 2 }),
+          })
+          .filter(({ vpcCidrMask, subnetSpecs }) => {
+            for (const subnetSpec of subnetSpecs) {
+              // Requite subnet to be smaller than VPC
+              if (subnetSpec.cidrMask! <= vpcCidrMask) {
```

Comment:
What does ! do here?

Reply:
It asserts that cidrMask is not undefined. The type of cidrMask is `number | undefined`, but we're always specifying a value for it in the setup, so using `!` is fine in a test ... which makes the numeric comparison easier to write rather than having to have a separate expression to prove to the compiller that it's not undefined.